File: poolAE.m
classdef poolAE
    properties
        lrate = 0.01;
        batchsz = 100;
        bigbatchsz = 10000;
        moment = 0;
        lambda = 1;
        max_moment = 0.9;
        W;
        prec;
    end
    
    methods
        function pool = poolAE(visibleSize, hiddenSize, useSingle)
            if useSingle
                pool.prec = 'single';
            else
                pool.prec = 'double';
            end
            
            pool.W = zeros(visibleSize, hiddenSize, pool.prec);
            r = randperm(visibleSize);
            for h=1:hiddenSize
                pool.W(r(h),h) = 1;
            end
        end
        
        function [cost,grad] = calc_cost(pool, X1, X2)           
            P1 = X1 * pool.W;
            P2 = X2 * pool.W;
            Y1 = P1 * pool.W';
            Y2 = P2 * pool.W';
            
            cost = mean(0.5*sum((X1 - Y1).^2,2) + 0.5*sum((X2 - Y2).^2,2) + ...
                pool.lambda*0.5*sum((P1 - P2).^2,2),1);
            
            grad = (Y1 - X1)' * P1 + X1' * ((Y1 - X1) * pool.W) + ...
                (Y2 - X2)' * P2 + X2' * ((Y2 - X2) * pool.W) + ...
                pool.lambda * (X1 - X2)' * (P1 - P2);
            
            grad = grad / size(X1,1);
        end
        
        function pool = train(pool, X1, X2, max_epoch, useGPU)
            data_sz = size(X1,1);
            pool.moment = 0;            
            speed = zeros(size(pool.W), pool.prec);
            
            if isa(pool.W, 'single') && isa(X1, 'double')
                X1 = single(X1);
                X2 = single(X2);
            end
            
            % shuffle train data
            r = randperm(data_sz);
            X1 = X1(r,:);
            X2 = X2(r,:);
            
            if useGPU
                pool.W = gpuArray(pool.W);
                speed = gpuArray(speed);
                %pool.W = gsingle(pool.W);
                %speed = gsingle(speed);
            else
                pool.bigbatchsz = data_sz;
            end
            
            tic;
            for epoch = 1:max_epoch
                if epoch == 5
                    pool.moment = pool.max_moment;
                end
                cost = 0;
                
                % when using GPU, only subset of train data can fit in
                % devide memory
                for nn = (1:data_sz / pool.bigbatchsz)
                    if useGPU
                        clear X1gpu X2gpu
                        X1gpu = gpuArray(X1(1+pool.bigbatchsz*(nn-1):pool.bigbatchsz*nn,:));
                        X2gpu = gpuArray(X2(1+pool.bigbatchsz*(nn-1):pool.bigbatchsz*nn,:));
                        %X1gpu = gsingle(X1(1+pool.bigbatchsz*(nn-1):pool.bigbatchsz*nn,:));
                        %X2gpu = gsingle(X2(1+pool.bigbatchsz*(nn-1):pool.bigbatchsz*nn,:));
                    end
                    for n = (1:pool.bigbatchsz / pool.batchsz)
                        if useGPU
                            X1_sub = X1gpu(1+pool.batchsz*(n-1):pool.batchsz*n,:);
                            X2_sub = X2gpu(1+pool.batchsz*(n-1):pool.batchsz*n,:);
                        else
                            X1_sub = X1(1+pool.batchsz*(n-1):pool.batchsz*n,:);
                            X2_sub = X2(1+pool.batchsz*(n-1):pool.batchsz*n,:);
                        end
                        
                        [c, grad] = pool.calc_cost(X1_sub, X2_sub);
                        cost = cost + c;
                        
                        speed = speed * pool.moment - grad;
                        pool.W = pool.W + pool.lrate * speed;
                        pool.W = pool.W .* (pool.W > 0); % keep positive
                    end
                end
                cost = cost * pool.batchsz / data_sz;
                fprintf(1,'epoch %d \t cost %f\n',epoch, cost);
            end
            toc;
            
            if useGPU
                clear X1gpu X2gpu;
                pool.W = gather(pool.W);
            end
        end
    end
end

File: poolAE.py
import time
import numpy as np
import cudamat as cm

class poolAE:
    lrate = 0.01
    batchsz = 100
    bigbatchsz = 10000
    moment = 0
    lam = 2
    max_moment = 0.9
    
    def __init__(self, _data_dim, _pool_dim):
        self.data_dim = _data_dim
        self.pool_dim = _pool_dim
        self.W = np.zeros((self.data_dim, self.pool_dim))
        for i in range(self.pool_dim):
            self.W[np.random.randint(0, self.data_dim), i] = 1

    def calc_cost(self, X1, X2):
        P1 = cm.dot(self.Wgpu.T, X1)
        P2 = cm.dot(self.Wgpu.T, X2)
        Y1 = cm.dot(self.Wgpu, P1)
        Y2 = cm.dot(self.Wgpu, P1)

        Y1.subtract(X1)
        Y2.subtract(X2)
    
        PD = cm.empty(P1.shape)
        XD = cm.empty(X1.shape)
        P1.subtract(P2, target = PD)
        X1.subtract(X2, target = XD)
    
        grad = cm.dot(XD, PD.T)
        grad.mult(self.lam)
        grad.add(cm.dot(Y1, P1.T))
        grad.add(cm.dot(Y2, P2.T))
        grad.add(cm.dot(X1, cm.dot(Y1.T, self.Wgpu)))
        grad.add(cm.dot(X2, cm.dot(Y2.T, self.Wgpu)))
        grad.divide(X1.shape[1])

        PD.mult(PD)
        Y1.mult(Y1)
        Y2.mult(Y2)
        cost = PD.sum(axis = 0)
        cost.mult(self.lam)
        cost.add(Y1.sum(axis = 0))
        cost.add(Y2.sum(axis = 0))
        cost = cost.sum(axis = 1)
        cost.divide(X1.shape[1] * 2)

        return [cost, grad]

    def train_step(self, X1, X2):
        [cost, grad] = self.calc_cost(X1, X2)
    
        grad.mult(self.lrate)
        self.speed.mult(self.moment)
        self.speed.subtract(grad)
        self.Wgpu.add(self.speed)
        Wmask = cm.empty(self.W.shape)
        self.Wgpu.greater_than(0, target = Wmask)
        self.Wgpu.mult(Wmask)
        
        return cost
        
    def train_init(self):
        # init cudamat
        cm.cublas_init()
        cm.CUDAMatrix.init_random(1)

        self.Wgpu = cm.CUDAMatrix(self.W)
        self.speed = cm.empty(self.W.shape)
        self.speed.assign(0)


    def train_finalize(self):
        self.Wgpu.copy_to_host()
        self.W = self.Wgpu.numpy_array
        print "CUDA try shutdown"
        cm.cublas_shutdown()
       

    def train(self, X1, X2, max_epoch):
        self.train_init()

        datasz = X1.shape[1]
 
        X1gpu = cm.CUDAMatrix(X1)
        X2gpu = cm.CUDAMatrix(X2)
        X1sub = cm.empty((self.data_dim, self.batchsz))
        X2sub = cm.empty((self.data_dim, self.batchsz))

        tic = time.time()
        for epoch in range(max_epoch):
            if epoch == 5:
                self.moment = 0.9

            for n in range(datasz / self.batchsz):
                X1gpu.get_col_slice(self.batchsz * n, self.batchsz * (n+1), target = X1sub)
                X2gpu.get_col_slice(self.batchsz * n, self.batchsz * (n+1), target = X2sub)
                cost = self.train_step(X1sub, X2sub)
                
            print "epoch %d cost %f" % (epoch, cost.asarray()[0,0])

        toc = time.time()
        print "time %s" % (toc - tic)
        self.train_finalize()

#p = poolAE(1000,400)
#p.lrate = 0.001
#p.bigbatchsz = 10000
#X1 = np.random.rand(1000,1000)        
#X2 = np.random.rand(1000,1000)
#p.train(X1,X2,10)
